File: C-Plus-Plus/math/magic_number.cpp
/*
A number is said to be Magic Number, when the recursive sum of the digits is 1.
For example 1729
    1+7+2+9=19
    1+9=10
    1+0=1
So, it comes out be 1. Hence, 1729 is a Magic Number.
*/

#include <bits/stdc++.h>
using namespace std;

// Function for finding Sum of digits of a number.
int sum_of_digits(int num)
{
    int sum = 0;
    while (num > 0) {
        int rem = num % 10;
        sum = sum + rem;
        num = num / 10;
    }
    return sum;
}

// Function to check whether the Number is Magic Number or Not.
bool is_magic(int num)
{
    while (num > 9) {
        num = sum_of_digits(num);
    }
    return num == 1;
}

int main()
{
    int num;
    cout << "Enter the Number to check whether it is Magic Number or Not:" << endl;
    cin >> num;
    if (is_magic(num))
        cout << num << " is Magic Number." << endl;
    else
        cout << num << " is Not Magic Number." << endl;
    return 0;
}

/*
Sample I/0

Input:
Enter the Number to check whether it is Magic Number or Not:
1234
Output:
1234 is Magic Number.

*/

File: Dart/maths/MagicNumber.dart
/*
A magic number is a number that reduces to 1 after a sequential operation wherein each step we
replace the original number by the sum of its digits until the sum reaches a single digit.
*/

import 'dart:io';

// Function to check if number is a Magic Number
bool isMagicNumber(int num) {
  int sum = num;

  // iterate until sum is reduces to single digit
  while(sum > 9){
    int copy = sum;
    sum = 0;

    // calculating sum of all  digits
    while(copy > 0){
      int digit = copy % 10;
      sum += digit;
      copy ~/= 10;
    }
  }
  
  if(sum == 1){
    return true;
  }
  return false;
}

void main() {
  print("Enter a number :");
  int num = int.parse(stdin.readLineSync()!);

  // Call function to check if number is a Magic number
  if (isMagicNumber(num)) {
    print("$num is a Magic Number");
  } else {
    print("$num is not a Magic Number");
  }
}

/**
Space Complexity O(1)
Time Complexity  O(nlog(n))

Sample input/output:

Enter a number :
532
532 is a Magic Number

Enter a number :
123
123 is not a Magic Number
*/

